File: typescript/src/modules/tts.ts
import { TTSOptions, TTSProvider, TTSProvidersResponse, TTSChunk, TTSChunkEvent } from '../types';
import { PermissionDeniedError } from './api';

export class TTSModule {
    private baseUrl: string;
    private appId: string;
    private appName: string;
    private apiKey?: string;

    constructor(realtimexUrl: string, appId: string, appName?: string, apiKey?: string) {
        this.baseUrl = realtimexUrl.replace(/\/$/, '');
        this.appId = appId;
        this.appName = appName || process.env.RTX_APP_NAME || 'Local App';
        this.apiKey = apiKey;
    }

    private get headers(): Record<string, string> {
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
        };
        if (this.apiKey) {
            headers['Authorization'] = `Bearer ${this.apiKey}`;
        }
        if (this.appId) {
            headers['x-app-id'] = this.appId;
        }
        return headers;
    }

    /**
     * Request a single permission from Electron via internal API
     */
    private async requestPermission(permission: string): Promise<boolean> {
        try {
            const response = await fetch(`${this.baseUrl}/api/local-apps/request-permission`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    app_id: this.appId,
                    app_name: this.appName,
                    permission,
                }),
            });
            const data = await response.json();
            return data.granted === true;
        } catch (error) {
            console.error('[SDK] Permission request failed:', error);
            return false;
        }
    }

    /**
     * Internal request wrapper that handles automatic permission prompts
     */
    private async request<T>(method: string, endpoint: string, body?: any, isStream = false): Promise<T> {
        const response = await fetch(`${this.baseUrl}${endpoint}`, {
            method,
            headers: this.headers,
            body: body ? JSON.stringify(body) : undefined,
        });

        if (!response.ok) {
            const data = await response.json();

            if (data.code === 'PERMISSION_REQUIRED') {
                const permission = data.permission || 'tts.generate';
                const granted = await this.requestPermission(permission);
                if (granted) {
                    return this.request<T>(method, endpoint, body, isStream);
                }
                throw new PermissionDeniedError(permission);
            }

            throw new Error(data.error || `Request failed: ${response.status}`);
        }

        if (isStream) {
            return response.body as unknown as T;
        }

        const contentType = response.headers.get('content-type');
        if (contentType && contentType.includes('application/json')) {
            return response.json() as unknown as T;
        }

        return response.arrayBuffer() as unknown as T;
    }

    /**
     * Generate speech from text (returns full buffer)
     * 
     * @example
     * ```ts
     * const buffer = await sdk.tts.speak("Hello world");
     * // Play buffer...
     * ```
     */
    async speak(text: string, options: TTSOptions = {}): Promise<ArrayBuffer> {
        return this.request<ArrayBuffer>('POST', '/sdk/tts', {
            text,
            ...options
        });
    }

    /**
     * Generate speech from text with streaming (yields decoded audio chunks)
     * Uses SSE internally but returns decoded ArrayBuffer chunks for easy playback.
     * 
     * @example
     * ```ts
     * for await (const chunk of sdk.tts.speakStream("Hello world")) {
     *   // chunk.audio is ArrayBuffer (already decoded!)
     *   const blob = new Blob([chunk.audio], { type: chunk.mimeType });
     *   const audio = new Audio(URL.createObjectURL(blob));
     *   await audio.play();
     * }
     * ```
     */
    async *speakStream(text: string, options: TTSOptions = {}): AsyncGenerator<TTSChunk> {
        const response = await fetch(`${this.baseUrl}/sdk/tts/stream`, {
            method: 'POST',
            headers: this.headers,
            body: JSON.stringify({ text, ...options }),
        });

        if (!response.ok) {
            const data = await response.json();

            if (data.code === 'PERMISSION_REQUIRED') {
                const permission = data.permission || 'tts.generate';
                const granted = await this.requestPermission(permission);
                if (granted) {
                    yield* this.speakStream(text, options);
                    return;
                }
                throw new PermissionDeniedError(permission);
            }

            throw new Error(data.error || `Streaming failed: ${response.status}`);
        }

        const reader = response.body?.getReader();
        if (!reader) throw new Error('No response body');

        const decoder = new TextDecoder();
        let buffer = '';
        let eventType = '';

        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });

                // Parse SSE events
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    const trimmedLine = line.trim();
                    if (!trimmedLine) continue;

                    if (trimmedLine.startsWith('event:')) {
                        eventType = trimmedLine.slice(6).trim();
                    } else if (trimmedLine.startsWith('data:')) {
                        const eventData = trimmedLine.slice(5).trim();

                        if (eventType === 'chunk' && eventData) {
                            try {
                                const parsed = JSON.parse(eventData);
                                // Decode base64 to ArrayBuffer
                                const binaryString = atob(parsed.audio);
                                const bytes = new Uint8Array(binaryString.length);
                                for (let i = 0; i < binaryString.length; i++) {
                                    bytes[i] = binaryString.charCodeAt(i);
                                }

                                yield {
                                    index: parsed.index,
                                    total: parsed.total,
                                    audio: bytes.buffer,
                                    mimeType: parsed.mimeType,
                                };
                            } catch (e) {
                                console.warn('[TTS SDK] Failed to parse chunk:', e);
                            }
                        } else if (eventType === 'error' && eventData) {
                            try {
                                const err = JSON.parse(eventData);
                                throw new Error(err.error || 'TTS streaming error');
                            } catch (e) {
                                if (e instanceof Error && e.message !== 'TTS streaming error') {
                                    throw e;
                                }
                            }
                        }
                        // info and done events are handled silently

                        eventType = '';
                    }
                }
            }
        } finally {
            reader.releaseLock();
        }
    }

    /**
     * List available TTS providers with configuration options
     */
    async listProviders(): Promise<TTSProvider[]> {
        const data = await this.request<TTSProvidersResponse>('GET', '/sdk/tts/providers');
        return data.providers || [];
    }
}

File: python/realtimex_sdk/tts.py
"""
TTS Module for RealtimeX SDK (Python)

Provides access to Text-to-Speech capabilities.
"""
import json
from typing import Any, AsyncIterator, Dict, List, Optional, Union

import httpx

from .api import PermissionDeniedError


class TTSModule:
    """
    TTS operations for RealtimeX SDK.
    
    Example:
        # Speak (Buffer)
        audio_bytes = await sdk.tts.speak("Hello world")
        
        # Speak (Stream)
        async for chunk in sdk.tts.speak_stream("Hello world"):
            # process chunk
            pass
    """
    
    def __init__(self, base_url: str, app_id: str, app_name: str = "Local App", api_key: Optional[str] = None):
        self._base_url = base_url.rstrip("/")
        self._app_id = app_id
        self._app_name = app_name
        self._api_key = api_key
    
    @property
    def _headers(self) -> Dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        if self._app_id:
            headers["x-app-id"] = self._app_id
        return headers
    
    async def _request_permission(self, permission: str) -> bool:
        """Request a single permission from Electron via internal API."""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self._base_url}/api/local-apps/request-permission",
                    json={
                        "app_id": self._app_id,
                        "app_name": self._app_name,
                        "permission": permission,
                    },
                    timeout=60.0 # Long timeout for user interaction
                )
                data = response.json()
                return data.get("granted", False)
        except Exception as e:
            print(f"[SDK] Permission request failed: {e}")
            return False

    async def _request(self, method: str, endpoint: str, **kwargs) -> Union[bytes, httpx.Response]:
        """Internal request wrapper that handles automatic permission prompts."""
        if httpx is None:
            raise ImportError("httpx is required for async operations")

        async with httpx.AsyncClient() as client:
            url = f"{self._base_url}{endpoint}"
            
            # For streaming, we need to return the response object to iterate over it
            stream = kwargs.pop('stream', False)
            
            if stream:
                 # For streaming we can't use the simple wrapper logic easily because we need to yield
                 # This is handled in speak_stream directly
                 pass

            response = await client.request(method, url, headers=self._headers, **kwargs)
            
            if response.status_code == 200:
                return response.content

            try:
                data = response.json()
                if data.get("code") == "PERMISSION_REQUIRED":
                    permission = data.get("permission", "tts.generate")
                    granted = await self._request_permission(permission)
                    if granted:
                        return await self._request(method, endpoint, **kwargs)
                    raise PermissionDeniedError(permission)
                
                raise Exception(data.get("error", f"Request failed: {response.status_code}"))
            except json.JSONDecodeError:
                response.raise_for_status()
                return response.content

    async def speak(
        self,
        text: str,
        voice: Optional[str] = None,
        model: Optional[str] = None,
        speed: Optional[float] = None,
        provider: Optional[str] = None,
        language: Optional[str] = None,
        num_inference_steps: Optional[int] = None
    ) -> bytes:
        """
        Generate speech from text (returns full audio bytes).
        
        Args:
            text: Text to speak
            voice: Optional voice ID
            model: Optional model ID
            speed: Optional speed multiplier
            provider: Optional provider ID
            language: Optional language code
            num_inference_steps: Optional quality (Supertonic)
            
        Returns:
            bytes: Audio data
        """
        payload = {
            "text": text,
            "voice": voice,
            "model": model,
            "speed": speed,
            "provider": provider,
            "language": language,
            "num_inference_steps": num_inference_steps
        }
        # Filter None values
        payload = {k: v for k, v in payload.items() if v is not None}
        
        return await self._request("POST", "/sdk/tts", json=payload, timeout=60.0)

    async def speak_stream(
        self,
        text: str,
        voice: Optional[str] = None,
        model: Optional[str] = None,
        speed: Optional[float] = None,
        provider: Optional[str] = None,
        language: Optional[str] = None,
        num_inference_steps: Optional[int] = None
    ) -> AsyncIterator[Dict[str, Any]]:
        """
        Generate speech from text (yields decoded audio chunks).
        Uses SSE streaming internally, decodes base64 audio for you.
        
        Args:
            text: Text to speak
            voice: Optional voice ID
            model: Optional model ID
            speed: Optional speed multiplier
            provider: Optional provider ID
            language: Optional language code
            num_inference_steps: Optional quality (Supertonic)
            
        Yields:
            dict: {"index": int, "total": int, "audio": bytes, "mimeType": str}
        """
        import base64
        
        if httpx is None:
            raise ImportError("httpx is required for async operations")
            
        payload = {
            "text": text,
            "voice": voice,
            "model": model,
            "speed": speed,
            "provider": provider,
            "language": language,
            "num_inference_steps": num_inference_steps
        }
        payload = {k: v for k, v in payload.items() if v is not None}
        
        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST", 
                f"{self._base_url}/sdk/tts/stream", 
                headers=self._headers, 
                json=payload,
                timeout=120.0
            ) as response:
                
                if response.status_code != 200:
                    content = await response.aread()
                    try:
                        data = json.loads(content)
                        if data.get("code") == "PERMISSION_REQUIRED":
                            permission = data.get("permission", "tts.generate")
                            granted = await self._request_permission(permission)
                            if granted:
                                async for chunk in self.speak_stream(text, voice, model, speed, provider, language, num_inference_steps):
                                    yield chunk
                                return
                            raise PermissionDeniedError(permission)
                        raise Exception(data.get("error", f"Stream failed: {response.status_code}"))
                    except json.JSONDecodeError:
                        response.raise_for_status()
                
                # Parse SSE events
                buffer = ""
                event_type = ""
                
                async for raw_chunk in response.aiter_text():
                    buffer += raw_chunk
                    
                    while "\n" in buffer:
                        line, buffer = buffer.split("\n", 1)
                        line = line.strip()
                        
                        if not line:
                            continue
                        
                        if line.startswith("event:"):
                            event_type = line[6:].strip()
                        elif line.startswith("data:"):
                            event_data = line[5:].strip()
                            
                            if event_type == "chunk" and event_data:
                                try:
                                    parsed = json.loads(event_data)
                                    # Decode base64 audio
                                    audio_bytes = base64.b64decode(parsed["audio"])
                                    yield {
                                        "index": parsed["index"],
                                        "total": parsed["total"],
                                        "audio": audio_bytes,
                                        "mimeType": parsed.get("mimeType", "audio/wav"),
                                    }
                                except Exception as e:
                                    print(f"[TTS SDK] Failed to parse chunk: {e}")
                            elif event_type == "error" and event_data:
                                try:
                                    err = json.loads(event_data)
                                    raise Exception(err.get("error", "TTS streaming error"))
                                except json.JSONDecodeError:
                                    pass
                            
                            event_type = ""


    async def list_providers(self) -> List[Dict[str, Any]]:
        """
        List available TTS providers.
        
        Returns:
            List[Dict]: List of provider objects
        """
        if httpx is None:
            raise ImportError("httpx is required for async operations")
            
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self._base_url}/sdk/tts/providers",
                headers=self._headers,
                timeout=10.0
            )
            
            if response.status_code != 200:
                data = response.json()
                if data.get("code") == "PERMISSION_REQUIRED":
                    permission = data.get("permission", "tts.generate")
                    granted = await self._request_permission(permission)
                    if granted:
                        return await self.list_providers()
                    raise PermissionDeniedError(permission)
                raise Exception(data.get("error", f"Request failed: {response.status_code}"))
                
            data = response.json()
            return data.get("providers", [])
